xds/src/main/java/io/grpc/xds/XdsNameResolver2.java

```diff
@@ -0,0 +1,378 @@
+            .build();
+    listener.onResult(result);
+  }
+
+  @SuppressWarnings("ModifyCollectionInEnhancedForLoop")  // ok for concurrent map
+  private class ConfigWatcherImpl implements ConfigWatcher {
+    @Override
+    public void onConfigChanged(ConfigUpdate update) {
+      Set<String> clusters = new HashSet<>();
+      for (Route route : update.getRoutes()) {
+        RouteAction action = route.getRouteAction();
+        if (action.getCluster() != null) {
+          clusters.add(action.getCluster());
+        } else if (action.getWeightedCluster() != null) {
+          for (ClusterWeight weighedCluster : action.getWeightedCluster()) {
+            clusters.add(weighedCluster.getName());
+          }
+        }
+      }
+      boolean receivedNewCluster = false;
+      for (String newCluster : clusters) {
+        if (clusterRefs.putIfAbsent(newCluster, new AtomicInteger(1)) == null) {
+          receivedNewCluster = true;
+        }
+      }
+      // Update service config to include newly added clusters.
+      if (receivedNewCluster) {
+        updateResolutionResult();
+      }
+      // Make newly added clusters selectable by config selector.
+      routes = update.getRoutes();
+      // Drops reference for deleted clusters, update service config to remove deleted clusters
+      // not in use.
+      boolean shouldUpdateResult = false;
+      for (Map.Entry<String, AtomicInteger> entry : clusterRefs.entrySet()) {
+        if (!clusters.contains(entry.getKey())) {
+          int count = entry.getValue().decrementAndGet();
+          if (count == 0) {
+            clusterRefs.remove(entry.getKey());
+            shouldUpdateResult = true;
+          }
+        }
+      }
+      if (shouldUpdateResult) {
+        updateResolutionResult();
+      }
+    }
+
+    @Override
+    public void onResourceDoesNotExist(String resourceName) {
+      logger.log(XdsLogLevel.INFO, "Resource {0} is unavailable", resourceName);
+      ConfigOrError parsedServiceConfig =
+          serviceConfigParser.parseServiceConfig(Collections.<String, Object>emptyMap());
+      ResolutionResult result =
+          ResolutionResult.newBuilder()
+              .setServiceConfig(parsedServiceConfig)
+              // let channel take action for no config selector
+              .build();
+      listener.onResult(result);
+    }
+
+    @Override
+    public void onError(Status error) {
+      logger.log(
+          XdsLogLevel.WARNING,
+          "Received error from xDS client {0}: {1}", xdsClient, error.getDescription());
+      listener.onError(error);
+    }
+  }
+
+  @VisibleForTesting
+  static Map<String, ?> generateServiceConfigWithMethodConfig(
+      String fullMethodName, long timeoutNano) {
+    int index = fullMethodName.lastIndexOf('/');
+    String serviceName = fullMethodName.substring(0, index);
+    String methodName = fullMethodName.substring(index + 1);
+    String timeout = timeoutNano / 1_000_000_000.0 + "s";
```

Comment:
What happened to the nanosecond portion? Was it just dropped on the floor?

Reply:
No, it's not. Actually the test case covers it. It's at [`generateServiceConfig_forMethodTimeoutConfig`](https://github.com/grpc/grpc-java/pull/7275/files#diff-2f70ae4b0fffca51cf8c1e25ac25f794R408-R420) where the input timeout is 1 second plus 1 nanosecond, the output is `1.000000001s`

Reply:
Oh, I missed this was stored into a string.